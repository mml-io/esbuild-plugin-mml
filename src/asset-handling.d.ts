import * as esbuild from "esbuild";
import { Asset } from "./types";
export declare function createAssetHandler(sourceRoot: string, outdir: string, assetDir: string, assetPrefix: string, assets: Asset[], build: esbuild.PluginBuild): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult | null>;
