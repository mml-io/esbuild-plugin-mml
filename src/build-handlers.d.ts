import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function createEntryPointResolver(discoveredEntryPoints: DiscoveredEntryPoints, build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult | null>;
export declare function createMMLImportResolver(build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult>;
export declare function createMMLRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createMMLTSXRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createTSXHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createWorldHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createHTMLHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
