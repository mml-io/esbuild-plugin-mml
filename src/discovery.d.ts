import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function doDiscoveryBuild(esbuildOptions: esbuild.BuildOptions): Promise<DiscoveredEntryPoints>;
