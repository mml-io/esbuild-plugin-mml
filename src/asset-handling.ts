import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { Asset } from "./types";

export function createAssetHandler(
  sourceRoot: string,
  outdir: string,
  assetDir: string,
  assetPrefix: string,
  assets: Asset[],
  build: esbuild.PluginBuild,
) {
  return async (
    args: esbuild.OnLoadArgs,
  ): Promise<esbuild.OnLoadResult | null> => {
    const extension = path.extname(args.path);

    // Check if this file should be handled as an asset
    const loader = build.initialOptions.loader?.[extension];
    if (loader !== "file") {
      return null;
    }

    // Generate asset file name based on the source path
    const relativePath = path.relative(sourceRoot, args.path);
    const fileName = relativePath.replace(/[/\\]/g, "_");

    // Build full asset URL
    const assetUrl = assetPrefix + fileName;

    // Determine output path
    const outputPath = assetDir
      ? path.join(outdir, assetDir, fileName)
      : path.join(outdir, fileName);

    // Copy the asset to the output directory
    await fsp.mkdir(path.dirname(outputPath), { recursive: true });
    await fsp.copyFile(args.path, outputPath);

    // Track the asset
    assets.push({
      fileName,
      url: assetUrl,
      path: outputPath,
    });

    return {
      contents: assetUrl,
      loader: "text",
    };
  };
}
