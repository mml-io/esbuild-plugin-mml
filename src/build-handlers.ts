import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { DiscoveredEntryPoints } from "./types";

export function createEntryPointResolver(
  discoveredEntryPoints: DiscoveredEntryPoints,
  build: esbuild.PluginBuild,
) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult | null> => {
    if (args.kind !== "entry-point" || args.pluginData === "mml-skip-resolve") {
      return null;
    }

    const resolvedPath = await build.resolve(args.path, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    const namespace = discoveredEntryPoints.tsx.has(resolvedPath.path)
      ? "mml-build-tsx"
      : discoveredEntryPoints.html.has(resolvedPath.path)
        ? "mml-build-html"
        : discoveredEntryPoints.world.has(resolvedPath.path)
          ? "mml-build-world"
          : null;

    return namespace
      ? {
          path: resolvedPath.path,
          namespace,
          pluginData: resolvedPath.pluginData as unknown,
        }
      : null;
  };
}

export function createMMLImportResolver(build: esbuild.PluginBuild) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult> => {
    const importPath = args.path.replace(/^mml:/, "");
    const resolvedPath = await build.resolve(importPath, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    if (resolvedPath.errors.length > 0) {
      return {
        errors: resolvedPath.errors.map((error) => ({
          ...error,
          text: `Failed to resolve ${importPath} from ${args.importer}. ${error.text}`,
        })),
      };
    }

    return {
      path: resolvedPath.path,
      namespace: resolvedPath.path.endsWith(".html")
        ? "mml-build-ref"
        : "mml-build-tsx-ref",
      pluginData: resolvedPath.pluginData as string,
    };
  };
}

function createRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
  isTSX = false,
) {
  return (args: esbuild.OnLoadArgs): esbuild.OnLoadResult => {
    const relativePath = path.relative(sourceRoot, args.path);
    let outputUrl = entryPointToOutputUrl[relativePath];

    if (!outputUrl) {
      const baseName = path.basename(relativePath, path.extname(relativePath));
      const dirPath = path.dirname(relativePath);
      const pathParts = [dirPath, baseName].filter(
        (part) => part && part !== ".",
      );

      outputUrl =
        pathParts
          .join("-")
          .replace(isTSX ? /[^a-zA-Z0-9\-_]/g : /[/\\]/g, "-") + ".html";

      if (!isTSX) {
        entryPointToOutputUrl[relativePath] = outputUrl;
      }
    }

    return {
      contents: documentPrefix + outputUrl,
      loader: "text",
      resolveDir: path.dirname(args.path),
      watchFiles: [args.path],
    };
  };
}

export function createMMLRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    false,
  );
}

export function createMMLTSXRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    true,
  );
}

async function loadFileContent(
  args: esbuild.OnLoadArgs,
  loader: "tsx" | "text",
) {
  const content = await fsp.readFile(args.path, "utf-8");
  return {
    contents: content,
    loader,
    resolveDir: path.dirname(args.path),
    watchFiles: [args.path],
  };
}

export const createTSXHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createWorldHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createHTMLHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "text");
