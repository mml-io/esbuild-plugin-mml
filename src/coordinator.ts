import * as esbuild from "esbuild";
import { makeResultProcessor } from "./results";
import { worldContext } from "./world";
import { documentContext } from "./documents";

export interface CoordinatorOptions {
  verbose: boolean;
  options: esbuild.BuildOptions;
  worlds: string[];
  documents: string[];
  assetDir: string;
  processor: ReturnType<typeof makeResultProcessor>;
  build: esbuild.PluginBuild["esbuild"];
}

function subsetOf<T>(set: Set<T>, subset: Set<T>): boolean {
  return Array.from(subset).every((v) => set.has(v));
}

function union<T>(s1: Set<T>, s2: Set<T>): Set<T> {
  const result = new Set<T>();
  s1.forEach((v) => result.add(v));
  s2.forEach((v) => result.add(v));
  return result;
}

export class Coordinator {
  private running = false;
  private building = true;
  private documents: Set<string>;
  private discoveredDocuments = new Set<string>();
  private docsCtx?: esbuild.BuildContext;
  private worldCtx?: esbuild.BuildContext;

  constructor(private options: CoordinatorOptions) {
    this.documents = new Set(options.documents);
  }

  public async start() {
    this.worldCtx = await worldContext({
      ...this.options,
      onEnd: (result, discoveredDocuments, importStubs) => {
        discoveredDocuments.forEach((d) => this.discoveredDocuments.add(d));
        this.options.processor.pushResult("world", result, importStubs);
      },
    });

    void this.worldCtx.watch();

    this.running = true;
    await this.tick(true);
  }

  async tick(firstTick = false) {
    if (firstTick || !subsetOf(this.documents, this.discoveredDocuments)) {
      this.documents = union(this.documents, this.discoveredDocuments);
      this.discoveredDocuments.clear();

      const oldDocsCtx = this.docsCtx;

      this.docsCtx = await documentContext({
        ...this.options,
        documents: Array.from(this.documents),
        onDiscoveredDocuments: (discoveredDocuments) => {
          discoveredDocuments.forEach((d) => this.discoveredDocuments.add(d));
        },
        onEnd: async (result, importStubs) => {
          this.options.processor.pushResult("document", result, importStubs);
          await this.options.processor.process();
          this.building = false;
        },
      });

      void this.docsCtx.watch();

      await oldDocsCtx?.dispose();
    }
    setImmediate(() => {
      if (this.running) void this.tick();
    });
  }

  async finish(): Promise<void> {
    while (this.building) {
      await new Promise(setImmediate);
    }
    await Promise.all([this.docsCtx?.dispose(), this.worldCtx?.dispose()]);
    this.running = false;
  }
}
