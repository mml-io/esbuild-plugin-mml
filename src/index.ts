export * from "./plugin";
export * from "./world-config";
