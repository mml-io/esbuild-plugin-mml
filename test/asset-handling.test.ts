import * as esbuild from "esbuild";
import { mml } from "../src";
import fsp from "node:fs/promises";
import path from "node:path";
import { waitForDispose, walk } from "./test-helpers";

describe("Asset Handling", () => {
  const testOutdir = "test/generated/asset-handling";

  beforeEach(async () => {
    await fsp.rm(testOutdir, { recursive: true, force: true });
  });

  it("should handle assets in subdirectories", async () => {
    const outdir = path.join(testOutdir, "nested-assets");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    // Check for build errors and fail the test if any exist
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that the .glb asset file was copied to the output directory
    const assetsDir = path.join(outdir, "assets");
    const files = await fsp.readdir(assetsDir);
    expect(files).toContain("test_src_duck.glb");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle different asset types beyond .glb files", async () => {
    const outdir = path.join(testOutdir, "different-assets");
    
    // Create test assets
    await fsp.mkdir(path.join(testOutdir, "src"), { recursive: true });
    await fsp.writeFile(path.join(testOutdir, "src", "image.png"), "fake png data");
    await fsp.writeFile(path.join(testOutdir, "src", "audio.mp3"), "fake mp3 data");
    await fsp.writeFile(path.join(testOutdir, "src", "data.json"), JSON.stringify({ test: "data" }));
    await fsp.writeFile(path.join(testOutdir, "src", "test.ts"), `
      import image from "./image.png";
      import audio from "./audio.mp3";
      import data from "./data.json";
      console.log(image, audio, data);
    `);

    const config = {
      outdir,
      entryPoints: [`mml:${testOutdir}/src/test.ts`],
      loader: {
        ".png": "file" as const,
        ".mp3": "file" as const,
        ".json": "file" as const,
      },
      plugins: [mml({
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that all asset files were copied
    const assetsDir = path.join(outdir, "assets");
    const files = await fsp.readdir(assetsDir);
    expect(files).toContain("test_generated_asset-handling_src_image.png");
    expect(files).toContain("test_generated_asset-handling_src_audio.mp3");
    expect(files).toContain("test_generated_asset-handling_src_data.json");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should generate correct asset URLs with custom assetPrefix", async () => {
    const outdir = path.join(testOutdir, "custom-prefix");
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        assetDir: "static",
        assetPrefix: "https://cdn.example.com/v2/",
      })],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that assets use the custom prefix in the built file
    const builtFile = path.join(outdir, "test-src-a.html");
    const content = await fsp.readFile(builtFile, "utf-8");
    expect(content).toContain("https://cdn.example.com/v2/test_src_duck.glb");

    // Check that the asset was copied to the correct directory
    const staticDir = path.join(outdir, "static");
    const files = await fsp.readdir(staticDir);
    expect(files).toContain("test_src_duck.glb");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should copy assets to correct output directory based on assetDir option", async () => {
    const outdir = path.join(testOutdir, "custom-assetdir");
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        assetDir: "media/files",
      })],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that assets were copied to the nested directory structure
    const mediaDir = path.join(outdir, "media", "files");
    const files = await fsp.readdir(mediaDir);
    expect(files).toContain("test_src_duck.glb");

    // Verify the manifest includes the correct path
    const manifestPath = path.join(outdir, "manifest.json");
    const manifest = JSON.parse(await fsp.readFile(manifestPath, "utf-8"));
    expect(manifest.assetNameToPath["test_src_duck.glb"]).toBe("media/files/test_src_duck.glb");

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });

  it("should handle duplicate asset names in different directories", async () => {
    const outdir = path.join(testOutdir, "duplicate-names");
    
    // Create test structure with duplicate asset names
    await fsp.mkdir(path.join(testOutdir, "src/folder1"), { recursive: true });
    await fsp.mkdir(path.join(testOutdir, "src/folder2"), { recursive: true });
    await fsp.writeFile(path.join(testOutdir, "src/folder1", "asset.png"), "folder1 asset data");
    await fsp.writeFile(path.join(testOutdir, "src/folder2", "asset.png"), "folder2 asset data");
    await fsp.writeFile(path.join(testOutdir, "src", "test.ts"), `
      import asset1 from "./folder1/asset.png";
      import asset2 from "./folder2/asset.png";
      console.log(asset1, asset2);
    `);

    const config = {
      outdir,
      entryPoints: [`mml:${testOutdir}/src/test.ts`],
      loader: {
        ".png": "file" as const,
      },
      plugins: [mml({
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that both assets were copied with unique names
    const assetsDir = path.join(outdir, "assets");
    const files = await fsp.readdir(assetsDir);
    
    // Should have both assets with path-based naming to avoid conflicts
    const folder1Asset = files.find(f => f.includes("folder1") && f.includes("asset.png"));
    const folder2Asset = files.find(f => f.includes("folder2") && f.includes("asset.png"));
    
    expect(folder1Asset).toBeDefined();
    expect(folder2Asset).toBeDefined();
    expect(folder1Asset).not.toBe(folder2Asset);

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });



  it("should handle assets with complex paths containing special characters", async () => {
    const outdir = path.join(testOutdir, "special-chars");
    await fsp.rm(outdir, { recursive: true, force: true });

    // Create a subdirectory with special characters in the name
    const specialDir = path.join("test/src", "special-assets");
    await fsp.mkdir(specialDir, { recursive: true });

    // Create an asset with a complex path
    const assetPath = path.join(specialDir, "special-asset.glb");
    await fsp.copyFile("test/src/duck.glb", assetPath);

    // Create a test file that imports this asset
    const testFile = path.join("test/src", "special-import.ts");
    await fsp.writeFile(testFile, `
      import specialAsset from "./special-assets/special-asset.glb";
      export default function() {
        return specialAsset;
      }
    `);

    try {
      const config = {
        outdir,
        entryPoints: [`mml:test/src/special-import.ts`],
        loader: {
          ".glb": "file" as const,
        },
        plugins: [mml({
          assetDir: "assets",
          assetPrefix: "/special/",
        })],
      };

      const result = await esbuild.build(config);

      if (result.errors.length > 0) {
        throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
      }

      await waitForDispose();

      // Check that the asset was copied with a normalized name
      const manifestPath = path.join(outdir, "manifest.json");
      const manifestContent = await fsp.readFile(manifestPath, "utf-8");
      const manifest = JSON.parse(manifestContent);

      expect(Object.keys(manifest.assetNameToPath)).toContain("test_src_special-assets_special-asset.glb");
      expect(manifest.assetNameToPath["test_src_special-assets_special-asset.glb"]).toBe("assets/test_src_special-assets_special-asset.glb");

      // Check that the asset file exists
      const assetOutputPath = path.join(outdir, "assets", "test_src_special-assets_special-asset.glb");
      await expect(fsp.access(assetOutputPath)).resolves.toBeUndefined();

      for await (const { path, content } of walk(outdir)) {
        expect(content).toMatchSnapshot(path);
      }
    } finally {
      // Clean up
      await fsp.rm(specialDir, { recursive: true, force: true });
      await fsp.rm(testFile, { force: true });
    }
  });
}); 