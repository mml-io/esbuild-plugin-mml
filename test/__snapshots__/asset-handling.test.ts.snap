// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/a.js 1`] = `
"(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/b.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-a.html": "test-src-a.html",
    "test-src-b.html": "test-src-b.html",
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {
    "test_src_duck.glb": "media/files/test_src_duck.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/media/files/test_src_duck.glb 1`] = `"test_src_duck.glb"`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/test-src-a.html 1`] = `
"<body></body><script>(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
</script>"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/test-src-b.html 1`] = `
"<body></body><script>(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
</script>"
`;

exports[`Asset Handling should copy assets to correct output directory based on assetDir option: generated/asset-handling/custom-assetdir/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/a.js 1`] = `
"(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "https://cdn.example.com/v2/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/b.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-a.html": "test-src-a.html",
    "test-src-b.html": "test-src-b.html",
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {
    "test_src_duck.glb": "static/test_src_duck.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "https://cdn.example.com/v2/"
}"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/static/test_src_duck.glb 1`] = `"test_src_duck.glb"`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/test-src-a.html 1`] = `
"<body></body><script>(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "https://cdn.example.com/v2/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
</script>"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/test-src-b.html 1`] = `
"<body></body><script>(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
</script>"
`;

exports[`Asset Handling should generate correct asset URLs with custom assetPrefix: generated/asset-handling/custom-prefix/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/a.js 1`] = `
"(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/assets/test_src_duck.glb 1`] = `"test_src_duck.glb"`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/b.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-a.html": "test-src-a.html",
    "test-src-b.html": "test-src-b.html",
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {
    "test_src_duck.glb": "assets/test_src_duck.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/test-src-a.html 1`] = `
"<body></body><script>(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
</script>"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/test-src-b.html 1`] = `
"<body></body><script>(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
</script>"
`;

exports[`Asset Handling should handle assets in subdirectories: generated/asset-handling/nested-assets/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;

exports[`Asset Handling should handle assets with complex paths containing special characters: generated/asset-handling/special-chars/assets/test_src_special-assets_special-asset.glb 1`] = `"test_src_special-assets_special-asset.glb"`;

exports[`Asset Handling should handle assets with complex paths containing special characters: generated/asset-handling/special-chars/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-special-import.html": "test-src-special-import.html"
  },
  "assetNameToPath": {
    "test_src_special-assets_special-asset.glb": "assets/test_src_special-assets_special-asset.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/special/"
}"
`;

exports[`Asset Handling should handle assets with complex paths containing special characters: generated/asset-handling/special-chars/special-import.js 1`] = `
"(() => {
  var special_asset_default = "/special/test_src_special-assets_special-asset.glb";
  function special_import_default() {
    return special_asset_default;
  }
})();
"
`;

exports[`Asset Handling should handle assets with complex paths containing special characters: generated/asset-handling/special-chars/test-src-special-import.html 1`] = `
"<body></body><script>(() => {
  var special_asset_default = "/special/test_src_special-assets_special-asset.glb";
  function special_import_default() {
    return special_asset_default;
  }
})();
</script>"
`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/assets/test_generated_asset-handling_src_audio.mp3 1`] = `"test_generated_asset-handling_src_audio.mp3"`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/assets/test_generated_asset-handling_src_data.json 1`] = `"{"test":"data"}"`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/assets/test_generated_asset-handling_src_image.png 1`] = `"test_generated_asset-handling_src_image.png"`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-generated-asset-handling-src-test.html": "test-generated-asset-handling-src-test.html"
  },
  "assetNameToPath": {
    "test_generated_asset-handling_src_audio.mp3": "assets/test_generated_asset-handling_src_audio.mp3",
    "test_generated_asset-handling_src_data.json": "assets/test_generated_asset-handling_src_data.json",
    "test_generated_asset-handling_src_image.png": "assets/test_generated_asset-handling_src_image.png"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/test.js 1`] = `
"(() => {
  var image_default = "/test_generated_asset-handling_src_image.png";
  var audio_default = "/test_generated_asset-handling_src_audio.mp3";
  var data_default = "/test_generated_asset-handling_src_data.json";
  console.log(image_default, audio_default, data_default);
})();
"
`;

exports[`Asset Handling should handle different asset types beyond .glb files: generated/asset-handling/different-assets/test-generated-asset-handling-src-test.html 1`] = `
"<body></body><script>(() => {
  var image_default = "/test_generated_asset-handling_src_image.png";
  var audio_default = "/test_generated_asset-handling_src_audio.mp3";
  var data_default = "/test_generated_asset-handling_src_data.json";
  console.log(image_default, audio_default, data_default);
})();
</script>"
`;

exports[`Asset Handling should handle duplicate asset names in different directories: generated/asset-handling/duplicate-names/assets/test_generated_asset-handling_src_folder1_asset.png 1`] = `"test_generated_asset-handling_src_folder1_asset.png"`;

exports[`Asset Handling should handle duplicate asset names in different directories: generated/asset-handling/duplicate-names/assets/test_generated_asset-handling_src_folder2_asset.png 1`] = `"test_generated_asset-handling_src_folder2_asset.png"`;

exports[`Asset Handling should handle duplicate asset names in different directories: generated/asset-handling/duplicate-names/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-generated-asset-handling-src-test.html": "test-generated-asset-handling-src-test.html"
  },
  "assetNameToPath": {
    "test_generated_asset-handling_src_folder1_asset.png": "assets/test_generated_asset-handling_src_folder1_asset.png",
    "test_generated_asset-handling_src_folder2_asset.png": "assets/test_generated_asset-handling_src_folder2_asset.png"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Asset Handling should handle duplicate asset names in different directories: generated/asset-handling/duplicate-names/test.js 1`] = `
"(() => {
  var asset_default = "/test_generated_asset-handling_src_folder1_asset.png";
  var asset_default2 = "/test_generated_asset-handling_src_folder2_asset.png";
  console.log(asset_default, asset_default2);
})();
"
`;

exports[`Asset Handling should handle duplicate asset names in different directories: generated/asset-handling/duplicate-names/test-generated-asset-handling-src-test.html 1`] = `
"<body></body><script>(() => {
  var asset_default = "/test_generated_asset-handling_src_folder1_asset.png";
  var asset_default2 = "/test_generated_asset-handling_src_folder2_asset.png";
  console.log(asset_default, asset_default2);
})();
</script>"
`;
